File: gaiag/baseline/component/Alarm.js
// Gaiag --- Guile in Asd In Asd in Guile.
//
// This file is part of Gaiag.
//
// Copyright © 2014 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2014 Rutger van Beusekom <rutger.van.beusekom@verum.com>
//
// Gaiag is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Gaiag is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Gaiag.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

component.Alarm = function() {
  this.States= {
    Disarmed: 0, Armed: 1, Triggered: 2, Disarming: 3
  };

  this.state = this.States.Disarmed;
  this.sounding = false;

  this.console = new interface.IConsole();
  this.sensor = new interface.ISensor();
  this.siren = new interface.ISiren();

  this.console.in.arm = function() {
    console.log('Alarm.console_arm');
    if(this.state === this.States.Disarmed) {
      {
        this.sensor.in.enable();
        this.state = this.States.Armed;
      }
    }
    else if(this.state === this.States.Armed) {
      assert (false);
    }
    else if(this.state === this.States.Disarming) {
      assert (false);
    }
    else if(this.state === this.States.Triggered) {
      assert (false);
    }
  }.bind(this);
  this.console.in.disarm = function() {
    console.log('Alarm.console_disarm');
    if(this.state === this.States.Disarmed) {
      assert (false);
    }
    else if(this.state === this.States.Armed) {
      {
        this.sensor.in.disable();
        this.state = this.States.Disarming;
      }
    }
    else if(this.state === this.States.Disarming) {
      assert (false);
    }
    else if(this.state === this.States.Triggered) {
      {
        this.sensor.in.disable();
        this.siren.in.turnoff();
        this.sounding = false;
        this.state = this.States.Disarming;
      }
    }
  }.bind(this);
  this.sensor.out.triggered = function() {
    console.log('Alarm.sensor_triggered');
    if(this.state === this.States.Disarmed) {
      assert (false);
    }
    else if(this.state === this.States.Armed) {
      {
        this.console.out.detected();
        this.siren.in.turnon();
        this.sounding = true;
        this.state = this.States.Triggered;
      }
    }
    else if(this.state === this.States.Disarming) {
      { }
    }
    else if(this.state === this.States.Triggered) {
      assert (false);
    }
  }.bind(this);
  this.sensor.out.disabled = function() {
    console.log('Alarm.sensor_disabled');
    if(this.state === this.States.Disarmed) {
      assert (false);
    }
    else if(this.state === this.States.Armed) {
      assert (false);
    }
    else if(this.state === this.States.Disarming) {
      {
        if(this.sounding) {
          this.console.out.deactivated();
          this.siren.in.turnoff();
          this.state = this.States.Disarmed;
          this.sounding = false;
        }
        else {
          this.console.out.deactivated();
          this.state = this.States.Disarmed;
        }
      }
    }
    else if(this.state === this.States.Triggered) {
      assert (false);
    }
  }.bind(this);

};

File: gaiag/baseline/component/Alarm.java
// Gaiag --- Guile in Asd In Asd in Guile.
//
// This file is part of Gaiag.
//
// Copyright © 2014 Jan Nieuwenhuizen <janneke@gnu.org>
//
// Gaiag is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Gaiag is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Gaiag.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

class Alarm{
  enum States {
    Disarmed, Armed, Triggered, Disarming
  };

  States state;
  Boolean sounding;

  IConsole console;
  ISensor sensor;
  ISiren siren;

  public Alarm() {
    console = new IConsole();
    sensor = new ISensor();
    siren = new ISiren();
    console.getIn().arm = new Action() {
      public void action() {
        console_arm() ;
      }
    } ;
    console.getIn().disarm = new Action() {
      public void action() {
        console_disarm() ;
      }
    } ;
    sensor.getOut().triggered = new Action() {
      public void action() {
        sensor_triggered() ;
      }
    } ;
    sensor.getOut().disabled = new Action() {
      public void action() {
        sensor_disabled() ;
      }
    } ;
    public void console_arm() {
      System.err.println("Alarm.console_arm");
      if (state == States.Disarmed) {
        {
          sensor.getIn().enable();
          state = States.Armed;
        }
      }
      else if (state == States.Armed) {
        assert(false);
      }
      else if (state == States.Disarming) {
        assert(false);
      }
      else if (state == States.Triggered) {
        assert(false);
      }
    };

    public void console_disarm() {
      System.err.println("Alarm.console_disarm");
      if (state == States.Disarmed) {
        assert(false);
      }
      else if (state == States.Armed) {
        {
          sensor.getIn().disable();
          state = States.Disarming;
        }
      }
      else if (state == States.Disarming) {
        assert(false);
      }
      else if (state == States.Triggered) {
        {
          sensor.getIn().disable();
          siren.getIn().turnoff();
          sounding = false;
          state = States.Disarming;
        }
      }
    };

    public void sensor_triggered() {
      System.err.println("Alarm.sensor_triggered");
      if (state == States.Disarmed) {
        assert(false);
      }
      else if (state == States.Armed) {
        {
          console.getOut().detected();
          siren.getIn().turnon();
          sounding = true;
          state = States.Triggered;
        }
      }
      else if (state == States.Disarming) {
        { }
      }
      else if (state == States.Triggered) {
        assert(false);
      }
    };

    public void sensor_disabled() {
      System.err.println("Alarm.sensor_disabled");
      if (state == States.Disarmed) {
        assert(false);
      }
      else if (state == States.Armed) {
        assert(false);
      }
      else if (state == States.Disarming) {
        {
          if (sounding) {
            console.getOut().deactivated();
            siren.getIn().turnoff();
            state = States.Disarmed;
            sounding = false;
          }
          else {
            console.getOut().deactivated();
            state = States.Disarmed;
          }
        }
      }
      else if (state == States.Triggered) {
        assert(false);
      }
    };

  };
}
